HeartbeatLogging/Tests/Unit/HeartbeatStorageTests.swift

```diff
@@ -17,178 +17,315 @@ import XCTest
     // Given
-    let coderFake = CoderFake()
-    let storage = HeartbeatStorage(id: #file, storage: StorageFake(), coder: coderFake)
-    XCTAssertNil(storage.flush())
-    coderFake.failOnNextEncode = true
+    XCTAssertEqual(HeartbeatStorage.cachedInstances, [:])
     // When
-    storage.offer(Heartbeat(info: #function))
+    let heartbeatStorage = HeartbeatStorage.getInstance(id: "sparky")
     // Then
-    XCTAssertNil(storage.flush())
+    XCTAssertEqual(
+      HeartbeatStorage.cachedInstances,
+      ["sparky": Weak(object: heartbeatStorage)]
+    )
+
+    addTeardownBlock {
+      // Assert that deallocated `HeartbeatStorage` is removed
+      // from cached instances.
+      XCTAssertEqual(HeartbeatStorage.cachedInstances, [:])
+    }
   }
 
-  func testOfferHeartbeatWithDecodingError() throws {
+  func testGettingInstance_WhenInstanceDoesExist_ReturnsExistingInstance() {
     // Given
-    let coderFake = CoderFake()
-    let storage = HeartbeatStorage(id: #file, storage: StorageFake(), coder: coderFake)
-    coderFake.failOnNextDecode = true
+    let cachedHeartbeatStorage = HeartbeatStorage.getInstance(id: "sparky")
+    XCTAssertEqual(
+      HeartbeatStorage.cachedInstances,
+      ["sparky": Weak(object: cachedHeartbeatStorage)]
+    )
     // When
-    storage.offer(Heartbeat(info: #function))
+    let heartbeatStorage = HeartbeatStorage.getInstance(id: "sparky")
     // Then
-    XCTAssertNil(storage.flush())
+    XCTAssert(heartbeatStorage === cachedHeartbeatStorage)
+
+    addTeardownBlock {
+      // Assert that deallocated `HeartbeatStorage` is removed
+      // from cached instances.
+      XCTAssertEqual(HeartbeatStorage.cachedInstances, [:])
+    }
   }
 
-  func testOfferHeartbeatWithErrorThenOffer() throws {
+  func testCachedInstancesAreRemovedUponDeinit() {
     // Given
-    let (storageFake, queue) = (StorageFake(), DispatchQueue(label: #function))
-    let storage = HeartbeatStorage(id: #file, storage: storageFake, queue: queue)
-    storageFake.failOnNextWrite = true
-    storage.offer(Heartbeat(info: #function))
-    drain(queue)
-    XCTAssertNil(storage.flush())
+    XCTAssertEqual(HeartbeatStorage.cachedInstances, [:])
+    var heartbeatStorage1: HeartbeatStorage? = .getInstance(id: "sparky")
+    var heartbeatStorage2: HeartbeatStorage? = .getInstance(id: "sparky")
+    XCTAssertNotNil(heartbeatStorage1)
+    XCTAssertNotNil(heartbeatStorage2)
+    XCTAssert(heartbeatStorage1 === heartbeatStorage2)
     // When
-    storage.offer(Heartbeat(info: #function))
+    heartbeatStorage1 = nil
+    // - Then
+    XCTAssertNil(heartbeatStorage1)
+    XCTAssertNotNil(heartbeatStorage2)
+    XCTAssertEqual(
+      HeartbeatStorage.cachedInstances,
+      ["sparky": Weak(object: heartbeatStorage2)]
+    )
+    // When
+    heartbeatStorage2 = nil
+    // - Then
+    XCTAssertNil(heartbeatStorage2)
+    XCTAssertEqual(HeartbeatStorage.cachedInstances, [:])
```

Comment:
Note that implementation details are used here which force the test to change when the implementation changes even without changing the public API. Asserting implementation details should be avoided when possible. Doesn't look like a big concern in this case, but it seems to be avoidable. Not a blocker for the PR.

Reply:
Hmm, thanks for pointing this out. I'd like to understand your point more. So in this case, are the exposed implementation details the fact that it is a `Dictionary<String:WeakContainer>`? 

So then `HeartbeatStorage.cachedInstances` should be private? 
Without checking backing dictionary, I'm not sure how to confirm that an instance is removed from `cachedInstances`... would checking that a new instance is returned be enough?

```swift
var heartbeatStorage: HeartbeatStorage? = HeartbeatStorage.getInstance(id: "sparky")
heartbeatStorage = nil

// Instead of inspecting dictionary contents, try to get Instance again and assert that it is not the same instance as before (implying that the old one was discarded and new one was created?).
let newHeartbeatStorage= HeartbeatStorage.getInstance(id: "sparky")
XCTAssert(newHeartbeatStorage !== heartbeatStorage) 
```

Reply:
Yes, this looks sufficient to validate the API contract.